File: Game Files/AIConnector.py
import DLFCN
import sys
sys.setdlopenflags(DLFCN.RTLD_NOW | DLFCN.RTLD_GLOBAL)
import lua
from utils import isNumber

# path to the main lua file
PATH_TO_LUA_MAIN = "train.lua"

# for lua interface
LG = lua.globals()

# loads the AI
LG.dofile(PATH_TO_LUA_MAIN)


# calls setParams in AIConnector.lua
# giving the dict as argument is possible, but it won't be a lua table then, which could cause trouble
#   therefore a serialized version is given and reconstructed in lua
# returns the set dict (serialized in lua style) (for debugging purposes mainly)
def setParams(params):
	cmd = "{"
	for key, value in params.items():
		if value is None:
			cmd += "{}=nil, ".format(key)
		elif isNumber(value):
			cmd += "{}={}, ".format(key, value) # no quotes needed for numbers
		else:
			cmd += "{}=\"{}\", ".format(key, value) # lua table constructor ignores trailing commas, so this is fine
	cmd += "}"
	LG.aiconnector.setParams(lua.eval(cmd))
	return cmd


# calls the getAction() method from AIConnector.lua and forwards its return value
def getAction():
	return LG.aiconnector.getAction()


# sets the global SCREENSHOT_FILEPATH in lua
def setScreenshotPath(path):
	LG.SCREENSHOT_FILEPATH = path
	return path # for debugging


# returns wether the AI requested a new level to be loaded (boolean)
def getReload():
	return LG.aiconnector.getReload()


# returns the action counter
def getActionCount():
	return LG.aiconnector.getActionCount()


# returns world count
def getWorldCount():
	return LG.aiconnector.getWorldCount()


# increases world count by given amount
def increaseWorldCount(amount=1):
	return LG.aiconnector.increaseWorldCount(amount)


# cleanup at program end (close open files etc)
def cleanup():
	LG.aiconnector.cleanup()

File: Game Files/AIConnector.lua
local aiconnector = {}

-- parameters
aiconnector.PARAMS = {}

-- files that have been opened
aiconnector.OPEN_FILES = {}


-- functions

-- main function which returns the action id
function aiconnector.getAction()
    -- call primary function
    on_frame_emulated()
    return actions.NEXT_ACTION
end


-- if the AI wants a new level, it will set PARAMS.reload to true, which is forwarded to python by this method
function aiconnector.getReload()
    return aiconnector.PARAMS.reload == true -- comparison will turn nil into false for python
end


-- returns the action counter from lua (total number of taken actions)
function aiconnector.getActionCount()
	return STATS.ACTION_COUNTER
end


-- returns the world count
function aiconnector.getWorldCount()
	return STATS.WORLD_COUNT
end


-- increases world count by one
function aiconnector.increaseWorldCount(amount)
	amount = amount or 1 -- defaulting to 1
	STATS.WORLD_COUNT = STATS.WORLD_COUNT + 1
end


-- sets PARAMS (only needed because a return value with LUA_PREFIX is necessary
function aiconnector.setParams(par)
    aiconnector.PARAMS = par
end


function aiconnector.cleanup()
    for _, value in ipairs(aiconnector.OPEN_FILES) do
		value:close()
	end

end

return aiconnector